src/Esprima/Utils/JavascriptTextWriter.cs

```diff
@@ -0,0 +1,375 @@
+﻿using System.Runtime.CompilerServices;
+using Esprima.Ast;
+
+namespace Esprima.Utils;
+
+public delegate object? NodePropertyValueAccessor(Node node);
+
+public delegate ref readonly NodeList<T> NodePropertyListValueAccessor<T>(Node node) where T : Node?;
+
+/// <summary>
+/// Base Javascript text writer (code formatter) which uses the most compact possible (i.e. minimal) format.
+/// </summary>
+public partial class JavascriptTextWriter
+{
+    public record class Options
+    {
+        public static readonly Options Default = new();
+    }
+
+    public delegate JavascriptTextWriter Factory(TextWriter writer, Options options);
+
+    private readonly TextWriter _writer;
+
+    public JavascriptTextWriter(TextWriter writer, Options options)
+    {
+        _writer = writer ?? throw new ArgumentNullException(nameof(writer));
```

Comment:
FYI, `EsprimaExceptionHelper` isn't used consistently throughout the code base.

As a matter of fact, I fail to see why it exists at all (when throw expressions are a thing since C# 7.0). What do we gain from e.g. `Description = description ?? ThrowArgumentNullException<string>(nameof(description));` over `Description = description ?? throw new ArgumentNullException(nameof(description));`? I mean it's not even shorter and you need to type out the generic parameter too. Plus it breaks static analysis. (Not even `DoesNotReturn` solves this completely as the compiler still complains when it's used in a switch case branch.)

So how about getting rid of `EsprimaExceptionHelper` instead? I'm willing to do that if you sign off on it.

and I suggest getting rid of it altogether.

Reply:
It's for JIT mostly, JIT cannot inline methods that have `throw` in them (call stack would be wrong).

Reply:
Wow, I learned something today again! Thanks for this valuable info. I'll revise `throw` usage and change it to the helper method where inlining justifies it.

Reply:
Yeah, method must of course be quite small to be inline candidate (that's why some usages would be more for consistency). Also I *think* that the return of `T` for some throw helpers might not work when wanting the "throw helper pattern", they might need to be void, but not an issue for this PR, just a mental note.

Reply:
I annotated all my methods with `[MethodImpl(MethodImplOptions.AggressiveInlining)]` which I think it's beneficial to inline. Don't really care whether JIT inlines the rest or not, so in those cases I just use plain `throw`s. As I see, this pattern is followed in other parts of the code (e.g. in `JavascriptParser`). So, are we good or do you insist on using the throw helpers everywhere?

Reply:
No need, re-evaluate if needed later.